From 46b3aa9b6d37c4afe3a719b8a39a6d6d42e93cf0 Mon Sep 17 00:00:00 2001
From: "Nahulanthran, Sanjeev" <sanjeev.nahulanthran@intel.com>
Date: Wed, 12 Sep 2018 13:34:00 +0800
Subject: [PATCH] Add ACPI ID Matching for wm8731 and TLV320AIC3107

This method replaces the need for the platform board files required
for the wolfram codec, wm8731, and the TI codecs, AIC3107M and
AIC3107S. The ACPI IDs here have to be matched with the IDs placed
into the modified acpi_override tables placed in the /boot folder of
the target platform. ACPI IDs used shown below:
	1. wm8731   - INT345A
	2. AIC3107M - INT345B
	3. AIC3107S - INT345C

Change-Id: I34212fcd02a8e10016d1b9f5c1de6797ab962091
Signed-off-by: Nahulanthran, Sanjeev <sanjeev.nahulanthran@intel.com>
Signed-off-by: Ho, Yu Xin <yu.xin.ho@intel.com>

Conflicts:
	sound/soc/intel/boards/apli_lhcrb_aic3107M.c
	sound/soc/intel/boards/apli_lhcrb_aic3107S.c
	sound/soc/intel/boards/apli_lhcrb_wm8731.c
---
 sound/soc/codecs/tlv320aic3x.c               | 18 +++++++++++++++++-
 sound/soc/codecs/wm8731.c                    |  9 +++++++++
 sound/soc/intel/boards/apli_lhcrb_aic3107M.c |  3 ++-
 sound/soc/intel/boards/apli_lhcrb_aic3107S.c |  2 +-
 sound/soc/intel/boards/apli_lhcrb_wm8731.c   |  3 ++-
 5 files changed, 31 insertions(+), 4 deletions(-)

diff --git a/sound/soc/codecs/tlv320aic3x.c b/sound/soc/codecs/tlv320aic3x.c
index 06f9257..1ada705 100644
--- a/sound/soc/codecs/tlv320aic3x.c
+++ b/sound/soc/codecs/tlv320aic3x.c
@@ -38,6 +38,7 @@
 #include <linux/delay.h>
 #include <linux/pm.h>
 #include <linux/i2c.h>
+#include <linux/acpi.h>
 #include <linux/gpio.h>
 #include <linux/regulator/consumer.h>
 #include <linux/of.h>
@@ -1767,6 +1768,14 @@ static void aic3x_configure_ocmv(struct i2c_client *client)
 	{ AIC3X_PAGE_SELECT, 0x00 },
 };
 
+const struct acpi_device_id aic3x_acpi_match[] = {
+	{"INT345B", AIC3X_MODEL_3007},
+	{"INT345C", AIC3X_MODEL_3007},
+	{},
+};
+MODULE_DEVICE_TABLE(acpi, aic3x_acpi_match);
+EXPORT_SYMBOL_GPL(aic3x_acpi_match);
+
 /*
  * If the i2c layer weren't so broken, we could pass this kind of data
  * around
@@ -1780,6 +1789,7 @@ static int aic3x_i2c_probe(struct i2c_client *i2c,
 	struct device_node *np = i2c->dev.of_node;
 	int ret, i;
 	u32 value;
+	const struct acpi_device_id *acpi_id;
 
 	aic3x = devm_kzalloc(&i2c->dev, sizeof(struct aic3x_priv), GFP_KERNEL);
 	if (!aic3x)
@@ -1839,7 +1849,12 @@ static int aic3x_i2c_probe(struct i2c_client *i2c,
 		aic3x->gpio_reset = -1;
 	}
 
-	aic3x->model = id->driver_data;
+	acpi_id = acpi_match_device(aic3x_acpi_match,&i2c->dev);
+	if(acpi_id) {
+		aic3x->model = acpi_id->driver_data;
+	} else {
+		aic3x->model = id->driver_data;
+	}
 
 	if (gpio_is_valid(aic3x->gpio_reset) &&
 	    !aic3x_is_shared_reset(aic3x)) {
@@ -1917,6 +1932,7 @@ static int aic3x_i2c_remove(struct i2c_client *client)
 	.driver = {
 		.name = "tlv320aic3x-codec",
 		.of_match_table = of_match_ptr(tlv320aic3x_of_match),
+		.acpi_match_table = ACPI_PTR(aic3x_acpi_match),
 	},
 	.probe	= aic3x_i2c_probe,
 	.remove = aic3x_i2c_remove,
diff --git a/sound/soc/codecs/wm8731.c b/sound/soc/codecs/wm8731.c
index 4f9a1eb..f58b5a6 100644
--- a/sound/soc/codecs/wm8731.c
+++ b/sound/soc/codecs/wm8731.c
@@ -19,6 +19,7 @@
 #include <linux/delay.h>
 #include <linux/pm.h>
 #include <linux/i2c.h>
+#include <linux/acpi.h>
 #include <linux/slab.h>
 #include <linux/regmap.h>
 #include <linux/regulator/consumer.h>
@@ -730,6 +731,13 @@ static int wm8731_spi_remove(struct spi_device *spi)
 };
 #endif /* CONFIG_SPI_MASTER */
 
+const struct acpi_device_id wm8731_acpi_match[] = {
+	{"INT345A", 0},
+	{},
+};
+MODULE_DEVICE_TABLE(acpi, wm8731_acpi_match);
+EXPORT_SYMBOL_GPL(wm8731_acpi_match);
+
 #if IS_ENABLED(CONFIG_I2C)
 static int wm8731_i2c_probe(struct i2c_client *i2c,
 			    const struct i2c_device_id *id)
@@ -801,6 +809,7 @@ static int wm8731_i2c_remove(struct i2c_client *client)
 	.driver = {
 		.name = "wm8731",
 		.of_match_table = wm8731_of_match,
+		.acpi_match_table = ACPI_PTR(wm8731_acpi_match),
 	},
 	.probe =    wm8731_i2c_probe,
 	.remove =   wm8731_i2c_remove,
diff --git a/sound/soc/intel/boards/apli_lhcrb_aic3107M.c b/sound/soc/intel/boards/apli_lhcrb_aic3107M.c
index 28cd9de..cabe471 100755
--- a/sound/soc/intel/boards/apli_lhcrb_aic3107M.c
+++ b/sound/soc/intel/boards/apli_lhcrb_aic3107M.c
@@ -22,6 +22,7 @@
 #include <sound/pcm.h>
 #include <sound/soc.h>
 #include <linux/gpio.h>
+#include <linux/acpi.h>
 #include <sound/pcm_params.h>
 #include "../../codecs/tlv320aic3x.h"
 
@@ -283,7 +284,7 @@ static int aic3107_be_hw_params_fixup(struct snd_soc_pcm_runtime *rtd,
 		.name = "SSP1-Codec",
 		.id = 0,
 		.cpu_dai_name = "SSP1 Pin",
-		.codec_name = "tlv320aic3x-codec.3-0018",
+		.codec_name = "i2c-INT345B:00",
 		.codec_dai_name = "tlv320aic3x-hifi",
 		.ignore_suspend = 1,
 		.ignore_pmdown_time = 1,
diff --git a/sound/soc/intel/boards/apli_lhcrb_aic3107S.c b/sound/soc/intel/boards/apli_lhcrb_aic3107S.c
index d088717..f3b7be1 100755
--- a/sound/soc/intel/boards/apli_lhcrb_aic3107S.c
+++ b/sound/soc/intel/boards/apli_lhcrb_aic3107S.c
@@ -288,7 +288,7 @@ static int aic3107_be_hw_params_fixup(struct snd_soc_pcm_runtime *rtd,
 		.name = "SSP1-Codec",
 		.id = 0,
 		.cpu_dai_name = "SSP1 Pin",
-		.codec_name = "tlv320aic3x-codec.3-0018",
+		.codec_name = "i2c-INT345C:00",
 		.codec_dai_name = "tlv320aic3x-hifi",
 		.ignore_suspend = 1,
 		.ignore_pmdown_time = 1,
diff --git a/sound/soc/intel/boards/apli_lhcrb_wm8731.c b/sound/soc/intel/boards/apli_lhcrb_wm8731.c
index 667d3d9..30b801c 100755
--- a/sound/soc/intel/boards/apli_lhcrb_wm8731.c
+++ b/sound/soc/intel/boards/apli_lhcrb_wm8731.c
@@ -23,6 +23,7 @@
 #include <sound/pcm.h>
 #include <sound/soc.h>
 #include <linux/gpio.h>
+#include <linux/acpi.h>
 #include <sound/pcm_params.h>
 #include "../../codecs/wm8731.h"
 
@@ -210,7 +211,7 @@ static int apli_wm8731_hw_params(struct snd_pcm_substream *substream,
 		.name = "SSP4-Codec",
 		.id = 1,
 		.cpu_dai_name = "SSP4 Pin",
-		.codec_name = "wm8731.3-001a",
+		.codec_name = "i2c-INT345A:00",
 		.codec_dai_name = "wm8731-hifi",
 		.ignore_suspend = 1,
 		.ignore_pmdown_time = 1,
-- 
1.9.1

